File: python/556_NextGreaterElement.py
class Solution:
    def nextGreaterElement(self, n: int) -> int:
        digits = deque()
        while n > 0:
            digits.appendleft(n % 10)
            n //= 10
        
        switched = False
        for i in range(len(digits) - 1, 0, -1):
            if digits[i - 1] < digits[i]:
                switched = True
                end = len(digits) - 1
                
                while digits[i - 1] >= digits[end]: end -= 1
                digits[i - 1], digits[end] = digits[end], digits[i - 1]
                
                j = len(digits) - 1
                while i < j:
                    digits[i], digits[j] = digits[j], digits[i]
                    i += 1
                    j -= 1
                break
        
        if not switched:
            return -1
        
        while digits:
            if n > 2147483647 / 10: return -1
            n *= 10
            n += digits.popleft();
        return n
            

File: java/556_NextGreaterElement.java
class Solution {
    public int nextGreaterElement(int n) {
        List<Integer> digits = new ArrayList<Integer>();
        while (n > 0) {
            digits.add(n % 10);
            n /= 10;
        }
        
        boolean switched = false;
        int sz = digits.size();
        for (int i = 0; i < sz - 1; ++i) {
            if (digits.get(i + 1) < digits.get(i)) {
                switched = true;
                
                // i+1 switch with beginning
                int temp = digits.get(i + 1);
                int begin = 0;
                while (digits.get(begin) == 0 || digits.get(begin) <= temp) begin++;
                
                digits.set(i + 1, digits.get(begin));
                digits.set(begin, temp);
                
                // reverse subarray 0 to i
                int j = 0;
                while (i > j) {
                    temp = digits.get(i);
                    digits.set(i--, digits.get(j));
                    digits.set(j++, temp);
                }
                break;
            }
        }
        
        if (!switched) return -1;
        while (!digits.isEmpty()) {
            if (n > Integer.MAX_VALUE / 10) return -1;
            n *= 10;
            n += digits.remove(digits.size() - 1);
        }
        return n;
    }
}
